### PR TITLE
feat: Implements diff accumulator using WrapperType

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR implements the diff accumulator using WrapperType and refactors the query crate.
- moves `mod_name/mod.rs` to `mod_name.rs`
- adds a LargestType associated type to LogicalPrimitiveType to get the largest type a logical primitive type can cast to
- removes LargestType from NativeType
- now common_function just use common_query's Error and Result
- renames `SubT` to `O` as it means output type
- `query::Error` isn't an opaque error anymore, as the `function-macro` depends on the inner error

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #555

### CODE REVIEW

File: src/common/function/src/scalars/aggregate/diff.rs [hunk 1]
```diff
@@ -22,40 +22,43 @@ use common_query::error::{
...
+use datatypes::types::{LogicalPrimitiveType, WrapperType};
 use datatypes::value::ListValue;
-use datatypes::vectors::{ConstantVector, ListVector};
+use datatypes::vectors::{ConstantVector, Helper, ListVector};
 use datatypes::with_match_primitive_type_id;
 use num_traits::AsPrimitive;
 use snafu::{ensure, OptionExt, ResultExt};
 
 // https://numpy.org/doc/stable/reference/generated/numpy.diff.html
+// T is the input type, O is the output type.
```
Comment (waynexia) [inline src/common/function/src/scalars/aggregate/diff.rs:33]:
```suggestion
// I is the input type, O is the output type.
```
for consistency

File: src/datatypes/src/types/date_type.rs [hunk 2]
```diff
@@ -59,6 +59,7 @@ impl LogicalPrimitiveType for DateType {
     type ArrowPrimitive = Date32Type;
     type Native = i32;
     type Wrapper = Date;
+    type LargestType = DateType;
```
Comment (waynexia) [inline src/datatypes/src/types/date_type.rs:62]:
```suggestion
    type LargestType = Self;
```
Would this be more clear?

Comment (waynexia) [reply src/datatypes/src/types/date_type.rs:62]: As well as other types that cannot up-cast

File: src/datatypes/src/types/primitive_type.rs [hunk 3]
```diff
@@ -29,35 +30,33 @@ use crate::value::{Value, ValueRef};
...
+impl_native_type!(f32);
+impl_native_type!(f64);
 
 /// Represents the wrapper type that wraps a native type using the `newtype pattern`,
 /// such as [Date](`common_time::Date`) is a wrapper type for the underlying native
 /// type `i32`.
 pub trait WrapperType:
     Copy
+    + Send
+    + Sync
```
Comment (waynexia) [inline src/datatypes/src/types/primitive_type.rs:58]: Why these?

Comment (evenyag) [reply src/datatypes/src/types/primitive_type.rs:58]: Accumulator should be send/sync, also these trait bounds are quite common, such as [ArrowNativeType](https://docs.rs/arrow/26.0.0/arrow/datatypes/trait.ArrowNativeType.html)

**Review by waynexia — APPROVED** [review]
lgtm

**Pull request merged**